File: analytics-sdk/analytics-storage-server/src/services/databaseService.js
const fs = require('fs');
const path = require('path');
const sqlite3 = require('sqlite3').verbose();
const ApiError = require('../errors/ApiError');

const DB_DIR = path.join(__dirname, '..', 'db');

// Ensure DB directory exists
if (!fs.existsSync(DB_DIR)) {
  fs.mkdirSync(DB_DIR);
}

// Helper to get DB instance
function getDB(dbName) {
  const dbPath = path.join(DB_DIR, `${dbName}.sqlite`);
  return new sqlite3.Database(dbPath);
}

// Create database file and table if not exists
async function createDatabaseAndTable(dbName, tableName) {
  const dbPath = path.join(DB_DIR, `${dbName}.sqlite`);

  const dbExists = fs.existsSync(dbPath);

  const db = getDB(dbName);

  return new Promise((resolve, reject) => {
    db.serialize(() => {
      if (!dbExists) {
        console.log(`Database ${dbName} created.`);
      }
      db.run(
        `CREATE TABLE IF NOT EXISTS ${tableName} (id TEXT PRIMARY KEY, data TEXT)`,
        (err) => {
          if (err) return reject(new ApiError(500, 'Failed to create table'));
          resolve();
        }
      );
    });
  });
}

// Insert data
async function insertData(dbName, tableName, data) {
  const db = getDB(dbName);

  const id = data.id;
  const jsonData = JSON.stringify(data);

  return new Promise((resolve, reject) => {
    db.run(
      `INSERT INTO ${tableName} (id, data) VALUES (?, ?)`,
      [id, jsonData],
      function (err) {
        if (err) return reject(new ApiError(500, 'Insert failed'));
        resolve();
      }
    );
  });
}

// Update data
async function updateData(dbName, tableName, data) {
  const db = getDB(dbName);

  const id = data.id;
  const jsonData = JSON.stringify(data);

  return new Promise((resolve, reject) => {
    db.run(
      `UPDATE ${tableName} SET data = ? WHERE id = ?`,
      [jsonData, id],
      function (err) {
        if (err) return reject(new ApiError(500, 'Update failed'));
        if (this.changes === 0) return reject(new ApiError(404, 'Record not found'));
        resolve();
      }
    );
  });
}

// Delete table
async function deleteTable(dbName, tableName) {
  const db = getDB(dbName);

  return new Promise((resolve, reject) => {
    db.run(`DROP TABLE IF EXISTS ${tableName}`, (err) => {
      if (err) return reject(new ApiError(500, 'Failed to delete table'));
      resolve();
    });
  });
}

module.exports = {
  createDatabaseAndTable,
  insertData,
  updateData,
  deleteTable,
};

File: analytics-sdk/analytics-storage-server/src/services/database_service.py
import os
import sqlite3
from src.errors.api_error import ApiError
from src.helpers.db_helper import check_and_create_table, insert_data, update_data, delete_data

DB_DIR = os.path.join(os.path.dirname(__file__), '..', 'db')

if not os.path.exists(DB_DIR):
    os.makedirs(DB_DIR)

def get_db(db_name):
    db_path = os.path.join(DB_DIR, f"{db_name}.sqlite")
    return sqlite3.connect(db_path)

async def create_database_and_table(db_name, table_name):
    db = get_db(db_name)
    try:
        await check_and_create_table(db_name, table_name)
    except Exception:
        raise ApiError(500, "Failed to create table")
    finally:
        db.close()

async def insert_data_service(db_name, table_name, data):
    db = get_db(db_name)
    try:
        await insert_data(db, table_name, data)
    except Exception:
        raise ApiError(500, "Insert failed")
    finally:
        db.close()

async def update_data_service(db_name, table_name, data):
    db = get_db(db_name)
    try:
        await update_data(db, table_name, data)
    except Exception:
        raise ApiError(500, "Update failed")
    finally:
        db.close()

async def delete_data_service(db_name, table_name, data_id):
    db = get_db(db_name)
    try:
        await delete_data(db, table_name, data_id)
    except Exception:
        raise ApiError(500, "Delete failed")
    finally:
        db.close()